lib/src/fixed_datetime_formatter.dart

```diff
@@ -0,0 +1,239 @@
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// A formatter and parser for [DateTime] in a fixed format [String] pattern.
+///
+/// For example, calling
+/// `FixedDateTimeCodec('YYYYMMDDhhmmss').decodeToLocal('19960425050322')` has
+/// the same result as calling `DateTime(1996, 4, 25, 5, 3, 22)`.
+///
+/// The allowed characters are
+/// * `Y`	a digit used in the time scale component “calendar year”
+/// * `M`	a digit used in the time scale component “calendar month”
+/// * `D`	a digit used in the time scale component “calendar day”
+/// * `E`	a digit used in the time scale component “decade”
+/// * `C`	a digit used in the time scale component “century”
+/// * `h`	a digit used in the time scale component “clock hour”
+/// * `m`	a digit used in the time scale component “clock minute”
+/// * `s`	a digit used in the time scale component “clock second”
+/// as specified in the ISO 8601 standard.
+///
+/// Non-allowed characters in the format [pattern] are ignored when decoding a
+/// string, in this case `YYYY kiwi MM` is the same format string as
+/// `YYYY------MM`. When encoding a datetime, the non-format characters are in
+/// the output verbatim.
+///
+/// Note: this class differs from [DateFormat] in that here, the characters are
+/// treated literally, i.e. the format string `YYY` matching `996` would result in
+/// the same as calling `DateTime(996)`. [DateFormat] on the other hand uses the
+/// specification in https://www.unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table,
+/// the format string (or "skeleton") `YYY` specifies only the padding, so
+/// `1996` would be a valid match. This limits it's use to format strings
+/// containing delimiters, as the parser would not know how many digits to take
+/// otherwise.
+class FixedDateTimeFormatter {
+  static final _validFormatCharacters = 'YMDEChms'.codeUnits;
+  static final yearCode = 'Y'.codeUnitAt(0);
+  static final monthCode = 'M'.codeUnitAt(0);
+  static final dayCode = 'D'.codeUnitAt(0);
+  static final decadeCode = 'E'.codeUnitAt(0);
+  static final centuryCode = 'C'.codeUnitAt(0);
+  static final hourCode = 'h'.codeUnitAt(0);
+  static final minuteCode = 'm'.codeUnitAt(0);
+  static final secondCode = 's'.codeUnitAt(0);
+
+  final String pattern;
+  final _blocks = _ParsedFormatBlocks();
+
+  /// Creates a new [FixedDateTimeFormatter] with the provided [pattern].
+  ///
+  /// The [pattern] interprets the characters mentioned in
+  /// [FixedDateTimeFormatter] to represent fields of a `DateTime` value. Other
+  /// characters are not special.
+  ///
+  /// There must at most be one sequence of each special character to ensure a
+  /// single source of truth when constructing the [DateTime], so a pattern of
+  /// `"CCCC-MM-DD, CC"` is invalid, because it has two separate `C` sequences.
+  FixedDateTimeFormatter(this.pattern) {
```

Comment:
I'd make `{bool this.isUtc = false}` (or `= true`) a class field too, and use it as the default for `decode`, which then takes `{bool? isUtc}` and uses the class field as default.

That way you can create a `FixedDateTimeFormatter` with a particular UTC behaivor, and use it at the `Codec` API (where `decode` has no arguments).

Reply:
I agree as it makes this very much aligned to the `Codec` interface, but would not use the optional parameter in the methods. This would only lead to confusing `Constructor(isUtc: false).method(isUtc: true)` statements.

Reply:
You *can* do that, but you won't.
In comparison, you can then do either of :
```dart
Constructor(isUtc: false).method()
// or
Constructor().method(isUtc: false);
```
We do that in other codecs (for `allowInvalid` IIRC).